_episodes_rmd/01-raster-structure.Rmd

```diff
@@ -1,406 +1,498 @@
+data before we read that data into R. It is ideal to do this before importing
+your data.
+
+```{r view-attributes-gdal}
+GDALinfo("data/NEON-DS-Airborne-Remote-Sensing/HARV/DSM/HARV_dsmCrop.tif")
+```
+
+If you wish to store this information in R, you can do the following:
+
+```{r 'capout'}
+HARV_dsmCrop_info <- capture.output(
+  GDALinfo("data/NEON-DS-Airborne-Remote-Sensing/HARV/DSM/HARV_dsmCrop.tif")
+)
+```
+
+Each line of text that was printed to the console is now stored as an element of
+the character vector `HARV_dsmCrop_info`.
+
+We will be exploring this data throughout this episode. By the end of this
+episode, you will be able to explain and understand the output above.
+
+## Raster Data in R
+
+Now that we've previewed the metadata for our `GeoTIFF`, let's import this
+raster dataset into `R` and explore its metadata more closely.
+
+## Open a Raster in R
+We can use the `raster()` function to open a raster in R.
+
+> ## Data Tip - Object names
+> To improve code
+> readability, file and object names should be used that make it clear what is in
+> the file. The data for this episode were collected over from Harvard Forest so
+> we'll use a naming convention of `datatype_HARV`.
+{: .callout}
+
+First we will load our raster file into `R` and view the data structure.
+
+```{r}
+DSM_HARV <- 
+  raster("data/NEON-DS-Airborne-Remote-Sensing/HARV/DSM/HARV_dsmCrop.tif")
+
+DSM_HARV
+```
+
+The information above includes a report of min and max values, but no other data
+range statistics. Similar to other R data structures like vectors and data frame
+columns, descriptive statistics for raster data can be retrieved like
+
+```{r 'rast_struct'}
+summary(DSM_HARV)
+```
+
+but note the warning - unless you force R to calculate these statistics using
+every cell in the raster, it will take a random sample of 10,000 cells and
+calculate from that instead. To force calculation on more, or even all values,
+you can try tuning the parameter `maxsamp`:
+
+```{r }
+# lets try all cells instead
+summary(DSM_HARV, maxsamp = ncell(DSM_HARV))
+```
+
+You may not see major differences in summary stats as `maxsamp` increases,
+except with very large rasters.
+
+To visualise this data in R using `ggplot2`, we need to convert it to a
+dataframe. We learned about dataframes in [an earlier
+lesson](https://datacarpentry.org/r-intro-geospatial/04-data-structures-part2/index.html).
+
+`raster` has an inbuilt method for conversion to a plottable `data.frame`.
+
+```{r}
+DSM_HARV_df <- raster::as.data.frame(DSM_HARV, xy = TRUE)
+```
+
+Now when we view the structure of our data, we will see a standard
+dataframe format.
+
+```{r}
+str(DSM_HARV_df)
+```
+
+We can use `ggplot()` to plot this data:
+
+```{r, fig.width= 7, fig.height=7}
+ggplot() +
+    geom_raster(data = DSM_HARV_df , aes(x = x, y = y, fill = HARV_dsmCrop)) +
+    scale_fill_viridis_c() +
+    coord_equal()
+```
+
+The add-on package `ggspatial` can also be used. The raster is still converted

```

Comment:
Let's move this to a callout box so that the main flow of the text only discusses one possibility.